Shuffle Individual Ocarina Notes

GSKirox opened the pull request:
New setting that prevent playing an Ocarina note until you find the corresponding item that unlocks it. This adds 5 new items in the game for each note : A, C up, C down, C left and C right. 
![image](https://user-images.githubusercontent.com/65768236/219085426-236abcb9-23f4-4b43-a379-336ebd96f362.png)

Three new models have been created to represent the 5 buttons : 

https://user-images.githubusercontent.com/65768236/219050789-c05860b9-699c-41b6-9e9d-5359e8594a4d.mp4

These models take the color of A button/C button defined in Cosmetics tab if the parameter "Item Model Colors Match Cosmetics" hae been checked. Else they keep their vanilla color.

The information of a note being unlocked is stored in the unused field of permanent flags for scene 0x50.

This setting forces the songs_as_items at true, else you softlock when trying to learn a song if you don't have the corresponding inputs. Interestingly enough, Ocarina minigame is not a softlock, it just cancels automatically after a few seconds if no input is given.


Logic : 
- can_play(song) has been updated to check if we have the corresponding notes of the song, even if the melodies themselves were randomized by the randomize_ocarina_melodies setting.
- Scarecrow song asks for 2 notes in logic.
- Lost woods Ocarina minigame asks for all 5 notes in logic, but you can cheese it with less if you're lucky.
- If the setting is not on, all 5 buttons are added to the world state, to simplify the logic (same behaviour as keysy)


Shoutouts and thanks (in no particular order) to : 
- TrenteR for the initial idea on Discord
- Maplestar for the item models
- Fenhl for handling the logic when using randomize_ocarina_melodies
- RealRob for fixing the asm hack and coloring the models
- Engineer for the audio decomp and brainstorm discussions on where to do the asm hack
- Everyone else who tested and/or took part in the discussions on this crazy idea

Update : added a visual cue in the menu to check which ocarina buttons you already collected.
If you hover on the Ocarina slot, a small rectangle will appear, showing the icons of the buttons you got.
![image](https://github.com/OoTRandomizer/OoT-Randomizer/assets/65768236/952d6b84-f93b-47f2-b3f8-2fd96a6958bb)

And added the buttons to the file select screen. If that setting is on, the layout will be slightly altered to add a third line below the songs and moving the death counter and SoA down : 
![image](https://github.com/OoTRandomizer/OoT-Randomizer/assets/65768236/e1d1abc3-d782-4e52-aa68-bbc2865ef46d)
If the setting is off, the file select will go back to its usual layout.

GSKirox committed:
Shuffle Individual Ocarina Notes

GSKirox committed:
Update presets

GSKirox committed:
CI.py fixes

fenhl commented:
I think the new setting could use a name that distinguishes it from the ocarina melodies setting a bit more clearly. Something like “Shuffle Ocarina Note Buttons”.

r0bd0g commented:
Why was one instance of "DAMPE" changed to "DAY" in ruleparser?

fenhl commented:
Copy/paste error.

fenhl commented on RuleParser.py:
```diff
@@ -481,7 +481,7 @@ def at_night(self, node):
             # tod has DAMPE or (tod == NONE and (ss or find a path from a provider))
             # parsing is better than constructing this expression by hand
-            return ast.parse("(tod & TimeOfDay.DAMPE) if tod else (state.has_all_of((Ocarina, Suns_Song)) or state.search.can_reach(spot.parent_region, age=age, tod=TimeOfDay.DAMPE))", mode='eval').body
+            return ast.parse("(tod & TimeOfDay.DAMPE) if tod else ((state.has(Ocarina) and state.can_play_song_note('Suns Song')) or state.search.can_reach(spot.parent_region, age=age, tod=TimeOfDay.DAY))", mode='eval').body
```
```suggestion
            return ast.parse("(tod & TimeOfDay.DAMPE) if tod else ((state.has(Ocarina) and state.can_play_song_note('Suns Song')) or state.search.can_reach(spot.parent_region, age=age, tod=TimeOfDay.DAMPE))", mode='eval').body
```

GSKirox replied:
Oh yeah that's a fat finger mistake

GSKirox committed:
Update RuleParser.py

Co-authored-by: Fenhl <fenhl@fenhl.net>

r0bd0g commented:
Do Pierre and Bonooru appear in the spoiler log? If so, I think current implementation would have them appear there before you could actually play the songs.

GSKirox commented:
I think you're right yeah, so the Bonooru event would need to be "is_child and can_play(Scarecrow_Song)" instead of "is_child and Ocarina"

GSKirox commented:
About the setting name, i'm not too attached to the current one and i'll gladly change it for clarity. 
“Shuffle Ocarina Note Buttons” works for me, any other opinion?

r0bd0g commented:
I'm don't think your proposed solution for Pierre/Bonooru actually works but I'm too lazy to look up exactly how it doesn't work right now.

GSKirox commented:
Hmm i don't see where the Bonooru event and the Pierre location are actually used in logic. Seems like all Scarecrow logic is going through either Scarecrow or Distant_Scarecrow, which checks for hook/longshot and can_play(Scarecrow Song)
Guess i'll test if they actually show up in the walkthrough tonight

rrealmuto commented:
Just FYI thie message IDs used by this hack should be updated after #1884

r0bd0g commented:
We really should draw a line as far as not accepting too many weird features like this one...
The button models look cool. I'm not convinced this will be fun to play, though. So I'm not sure how I feel about it.

GSKirox commented:
> Just FYI thie message IDs used by this hack should be updated after #1884
Right ! Thanks for heads up.

> We really should draw a line as far as not accepting too many weird features like this one... The button models look cool. I'm not convinced this will be fun to play, though. So I'm not sure how I feel about it. Can't be worse than bonks, tbh.
Maybe it's time for the Meme Tab idea to be brought up again 👀

r0bd0g commented:
You should find a way to add the collected buttons to the file select item info.
What happens when you try to do a song playback on the pause screen?

GSKirox committed:
Update Bombchu bag message, fix Bonooru event and day/night logic

GSKirox commented:
I've updated the message ID to be coherent with #1884.
Fixed an error in at_day/at_night where i removed the Ocarina check.
And i did some plando tests, the event that shows in the walkthrough is : 
"Pierre":                                                  "Scarecrow Song"
which logic is : 
"Pierre": "is_adult and Bonooru and not free_scarecrow",
So i added have_all_notes_for_song(Scarecrow_Song) for the Bonooru event, which ensures that we have at least 2 ocarina buttons before Pierre will show in the walkthrough.

GSKirox commented:
> You should find a way to add the collected buttons to the file select item info. What happens when you try to do a song playback on the pause screen?
Yeah i agree, i'll see if what Jaybone is doing works for this.
For the song playback on pause screen, the first automatical play by the game works like normal, and then when you replay it, the notes you don't have don't do any input. You can move out the song or press start to exit pause screen fine.

r0bd0g commented:
Why was one instance of "DAMPE" changed to "DAY" in ruleparser?

GSKirox commented:
Because i'm dumb and i pushed the same mistake Fenhl fixed earlier 🚶

Cuphat commented on Cosmetics.py:
```diff
@@ -705,6 +705,10 @@ def patch_button_colors(rom, settings, log, symbols):
 
             log_dict['colors'][patch] = color_to_hex(colors[patch])
 
+    if settings.correct_model_colors:
+        symbol = rom.sym('CORRECT_MODEL_COLORS')
+        rom.write_byte(symbol, 0x01)
+
```
You'll need to move this symbol into the cosmetic context in config.asm and increment the cosmetic version/add a new `patch_set` for the new cosmetic version. Otherwise, this will write garbage into the wrong location when using patch files of a differing version.

GSKirox committed:
Move Correct Model Colors to CFG

GSKirox commented:
Rebased on latest dev and moved CORRECT_MODEL_COLORS to the cosmetics context.

GSKirox committed:
Fix various conflicts with latest dev branch

GSKirox commented:
Rebasing intensifies SMILERS

fenhl commented:
Thanks for the rebase. I have now merged this into [my main branch](https://github.com/fenhl/OoT-Randomizer) for testing, minus the cosmetics changes due to a limitation of the cosmetics versioning system. While merging, I noticed that you changed the internal name of the existing ocarina melodies setting, not just the display name. I'd suggest keeping the existing internal name for compatibility. On that note, I still think that “Shuffle Ocarina Note Buttons” would be a better name for the new setting.

fenhl commented on Messages.py:
```diff
@@ -264,7 +264,12 @@
     (0x9019, "\x08\x13\x09You found a \x05\x41Bombchu Bag\x05\x40!\x01It has some \x05\x41Bombchus\x05\x40 inside!\x01Find more in tall grass."),
-    (0x901A, "\x08You can't buy Bombchus without a\x01\x05\x41Bombchu Bag\x05\x40!")
+    (0x901A, "\x08You can't buy Bombchus without a\x01\x05\x41Bombchu Bag\x05\x40!"),
+    (0x908B, "\x08You got the\x01\x05\x41Ocarina A Button!\x05\x40\x01You can now play \x9F on the Ocarina!"),
```
Message ID 0x908b is already used by chest game shuffle.

rrealmuto replied:
Seeds should no longer ~~generate~~ patch if a duplicate message ID is used. That was part of the silver rupee PR.

Just added a new unit test for this #1977

fenhl replied:
I already have this unit test on my branch. It did not catch this conflict because chest game shuffle patches messages using special code, not the messages lists.

rrealmuto replied:
That should probably be changed in a future update. At very least have a placeholder in messages.py or something.

cjohnson57 commented:
As far as I can tell this has been tested extensively on some branches, so I think we're pretty much good to go on it once conflicts are resolved and I review

GSKirox committed:
Fix conflicts + various bugs + ingame menu for buttons

GSKirox commented:
All conflicts should be resolved, including the futures ones that would have clashed with #1972
I've used ids following the new key models, and increased the EXTENDED_OBJECT_TABLE size to what's needed to support all keys + buttons.

Also took the occasion to rollback the setting rename change for the "Ocarina songs" setting, as per Fenhl's suggestion, to not break plandos.

And also added a new visual cue in the menu so you can check which buttons you already got.

cjohnson57 commented:
Any idea what that triforce piece failure in the CI is about?

GSKirox commented:
Looks like the shuffle notes + randomize melodies combo doesn't work anymore and causes logic errors.
Haven't figured out why yet :(

r0bd0g commented:
Don't accidentally revert logic fixes to carpet guy and others.
